primitives/src/crh/poseidon/mod.rs

```diff
@@ -1,427 +1,213 @@
-            let rc = P::ROUND_CST[*round_cst_idx];
 
-            // go over each of the state vectors and add the constant
-            for k in 0..vec_state.len() {
-                vec_state[k][j] += &rc;
-            }
-            *round_cst_idx += 1;
-        }
+        let mut instance = Self {
+            state,
+            pending: Vec::with_capacity(P::R),
+            input_size: constant_size,
+            updates_ctr: 0,
+            mod_rate,
+            _parameters: PhantomData,
+        };
 
-        // Apply the S-BOX to each of the elements of the state vector
-        // Use Montgomery simultaneous inversion
-        let mut w: Vec<P::Fr> = Vec::new();
-        let mut accum_prod = P::Fr::one();
+        // If personalization Vec is not multiple of the rate, we pad it with zero field elements.
+        // This will allow eventually to precompute the constants of the initial state. This
+        // is exactly as doing H(personalization, padding, ...). NOTE: this way of personalizing
+        // the hash is not mentioned in https://eprint.iacr.org/2019/458.pdf
+        if personalization.is_some(){
+            let personalization = personalization.unwrap();
 
-        w.push(accum_prod);
-        // Calculate the intermediate partial products
-        for i in 0..vec_state.len() {
-            for j in 0..P::T {
-                accum_prod = accum_prod * &vec_state[i][j];
-                w.push(accum_prod);
+            for &p in personalization.into_iter(){
+                instance.update(p);
             }
-        }
 
-        // if the accum_prod is zero, it means that one of the S-Boxes is zero
-        // in that case compute the inverses individually
-        if accum_prod == P::Fr::zero() {
-            for i in 0..vec_state.len() {
-                for j in 0..P::T {
-                    if vec_state[i][j] != P::Fr::zero() {
-                        vec_state[i][j] = vec_state[i][j].inverse().unwrap();
-                    }
-                }
-            }
-        } else {
-
-            // Calculate the inversion of the products
-            // The inverse always exists in this case
-            let mut w_bar = accum_prod.inverse().unwrap();
-
-            // Extract the individual inversions
-            let mut idx: i64 = w.len() as i64 - 2;
-            for i in (0..vec_state.len()).rev() {
-                for j in (0..P::T).rev() {
-                    let vec_1 = vec_state[i][j].clone();
-                    vec_state[i][j] = w_bar * &w[idx as usize];
-                    w_bar = w_bar * &vec_1;
-                    idx -= 1;
-                }
-            }
+            let padding = if personalization.len() % P::R != 0 {
+                P::R - ( personalization.len() % P::R )
+            } else {
+                0
+            };
 
+            for _ in 0..padding {
+                instance.update(F::zero());
+            }
```

Comment:
This way of padding allows collisions in the initial state.
Although probably not security critical for us, that behaviour is definitely not wanted.